### DISCUSSION TITLE
Dual I2S Output With Different Sample Rates

### DISCUSSION THREAD

**Discussion by LsHallo** [pschatzmann/ESP32-A2DP#330] [Q&A] [answered]
Hello,
I'm trying to achieve something unusual that I could not find done on the interwebs.
I'm trying to receive data via BT using this library and then send the data out via the two I2S ports of my ESP32.
Both I2S ports will have different sample rates (44.1 and 22.05kHz) and the 22.05kHz port will also only be mono audio instead of stereo. I've disabled the I2S output of this library and I am using the set_stream_reader callback to write data to the 2 I2S ports.

I'm currently using the following code to achieve what I'm trying to do.
Both outputs individually seem to work fine. I2S0 is also generating MCLK for my DAC board.
But as soon as I connect the second I2S device I2S0 will have weird output distortions. You can barely make out what should be coming out of the port. As soon as i disconnect the DAC from I2S1 the sound on I2S0 returns to normal. While I2S0 is distorted I2S1 sounds fine.  
The distortion can best be described as crackling and a little bit of white noise.

To summarize:
I2S0 only -> fine
I2S1 only -> fine
I2S0 + 1 -> I2S0 is distorted, I2S1 is fine

Things I tried:
- Use 44.1kHz for both outputs.
- Use 16 bit outputs
- Use stereo for both outputs
- Don't use MCLK for I2S0
- Different DMA buffer sizes 4x64, 4x256, 8x256, 4x1024
- Swap port numbers in code (PORT_DAC = I2S_NUM_1, PORT_WLED = I2S_NUM_0) (PORT_DAC still breaks for some reason).

A few questions:
- Is this even possible on ESP32?
    - I could not find anything explicitly stating that 2 individual outputs are not possible or that the sample rates have to be tied in ESP tech docs.
    - They seem to get their master clk from the same source and there are dividers set based on sample rate.
- Would this work better if I had a dedicated converter ESP that has one I2S input and one I2S output with reduced sample rate? (I don't have any more to test right now. Should be arriving by Thursday)

THE CODE:
```cpp
/* Double Output -> Double Trouble */
/* This might work someday. Currently it is buggy AF */

#include <Arduino.h>
#include "driver/i2s.h"
#include <soc/i2s_reg.h>
#include "esp_intr_alloc.h"
#include "BluetoothA2DPSink.h"
#include "SoundData.h"

// Full output port
static const i2s_port_t PORT_DAC = I2S_NUM_0;
static const i2s_pin_config_t PINS_DAC = {
    .bck_io_num = 18,
    .ws_io_num = 19,
    .data_out_num = 21,
    .data_in_num = I2S_PIN_NO_CHANGE
};
static const i2s_config_t I2S_CONFIG_DAC = {
    .mode = i2s_mode_t(I2S_MODE_MASTER | I2S_MODE_TX),
    .sample_rate = 44100,
    .bits_per_sample = I2S_BITS_PER_SAMPLE_32BIT,
    .channel_format = I2S_CHANNEL_FMT_RIGHT_LEFT,
    .communication_format = I2S_COMM_FORMAT_STAND_I2S,
    .intr_alloc_flags = ESP_INTR_FLAG_LEVEL1,
    .dma_buf_count = 4,
    .dma_buf_len = 1024,
    .use_apll = true,
    //.tx_desc_auto_clear = true
};

// Limited output (mono + 22.05kHz)
static const i2s_port_t PORT_WLED = I2S_NUM_1;
static const i2s_pin_config_t PINS_WLED = {
    .bck_io_num = 22,
    .ws_io_num = 25,
    .data_out_num = 27,
    .data_in_num = I2S_PIN_NO_CHANGE
};
static const i2s_config_t I2S_CONFIG_WLED = {
    .mode = i2s_mode_t(I2S_MODE_MASTER | I2S_MODE_TX),
    .sample_rate = 22050,
    .bits_per_sample = I2S_BITS_PER_SAMPLE_32BIT,
    .channel_format = I2S_CHANNEL_FMT_ONLY_LEFT,
    .communication_format = I2S_COMM_FORMAT_STAND_I2S,
    .intr_alloc_flags = ESP_INTR_FLAG_LEVEL1,
    .dma_buf_count = 4,
    .dma_buf_len = 768,
    //.use_apll = true,
    //.tx_desc_auto_clear = true
};

BluetoothA2DPSink a2dp_sink;

void data_received_callback(const uint8_t *data, uint32_t length) {
    int bytesWritten;
    esp_err_t err;

    Frame *samples = (Frame *) data;

    err = i2s_write_expand(PORT_DAC, data, length, 16, 32, (size_t*)&bytesWritten, portMAX_DELAY);
    if(err != ESP_OK) {
        Serial.println("Err: Could not write data to DAC I2S!");
    }

    // Super crude downsampling for second port
    for(uint32_t i = 0; i < length / 4; i++) {
        // int16_t leftSample = samples[i * 4].channel1;
        // int16_t rightSample = samples[i * 4].channel2;

        // samples[i] = (leftSample + rightSample) / 2;
        samples[i] = samples[i * 4];
    }

    err = i2s_write_expand(PORT_WLED, data, length / 4, 16, 32, (size_t*)&bytesWritten, portMAX_DELAY);
    if(err != ESP_OK) {
        Serial.println("Err: Could not write data to WLED I2S!");
    }
}

///////////
// SETUP //
///////////
void setup() {
    // Serial.begin(115200);
    esp_err_t err;

    // Just in case ;)
    i2s_driver_uninstall(I2S_NUM_0);
    i2s_driver_uninstall(I2S_NUM_1);

    // I2S drivers
    err = i2s_driver_install(PORT_DAC, &I2S_CONFIG_DAC, 0, NULL);
    if(err != ESP_OK) {
        Serial.println("Setup of DAC I2S failed!");
    }
    err = i2s_driver_install(PORT_WLED, &I2S_CONFIG_WLED, 0, NULL);
    if(err != ESP_OK) {
        Serial.println("Setup of WLED I2S failed!");
    }

    // Pin config
    err = i2s_set_pin(PORT_DAC, &PINS_DAC);
    if(err != ESP_OK) {
        Serial.println("Setup of DAC I2S pins failed!");
    }
    err = i2s_set_pin(PORT_WLED, &PINS_WLED);
    if(err != ESP_OK) {
        Serial.println("Setup of WLED I2S pins failed!");
    }

    // I2S0 MCLK -> GPIO0
	REG_WRITE(PIN_CTRL, 0xFF0); 
	PIN_FUNC_SELECT(PERIPHS_IO_MUX_GPIO0_U, FUNC_GPIO0_CLK_OUT1);

    a2dp_sink.set_stream_reader(data_received_callback, false);
    a2dp_sink.start("Party", true);
}

//////////
// LOOP //
//////////
void loop() {
    delay(50);
}
```

**Comment by pschatzmann** [comment, marked as answer]
I tend to think that it is a hardware problem and you don't provide enugh power for 2 output devices, I therefore suggest you try to power the two DACs from a separate source.

**Comment by LsHallo** [comment]
I really thought you were onto something. But sadly the issue persists. I'm now using a 5v 4A supply to power the ESP and dac on i2s1 and a 12v 3a power supply to power the dac on i2s0. The same distortion still happens.

**Comment by pschatzmann** [comment]
If it is not hardware, then I would suggest to try 2 different ESP32 implementations: 2.0.4 and 1.0.6.
The second thing which might cause this is your logic length / 4 but only sample rate /2. Did you try with half but stereo ?  This would explain your described behavor as well.

**Comment by LsHallo** [comment]
You were onto something. I needed to divide the samples by 2 since I'm using Frames now that are 32bit. Once I'd done that it worked while powering the ESP from USB with serial monitor, dac0 from 12v and dac1 from 5v external.
After i disconnected USB and also connected the ESP to the 5v supply everything broke again.
Then I tried to add another 5v supply just for the ESP in the form of my phone charger. This also doesn't work. I don't know what's going on but it seems to be hardware related. The soft part worked and hardware kinda worked.

I now also tried to memcpy the data and send the copy to the second dac. This also didn't help.

**Reply by pschatzmann** [reply]
So, I suggest to try to increase the buffer count now...

**Reply by LsHallo** [reply]
It's even weirder. The output only works fine if I have the serial monitor active. As soon as I close it the static is back. What is this!? :D
I've increased the buffers to 8x1024 for both outputs.

**Reply by LsHallo** [reply]
Okay I've changed MCLK from GPIO0 to PERIPHS_IO_MUX_U0TXD_U which is labeled RX on my board for some reason. Now it seems to work even without serial monitor.

**Reply by pschatzmann** [reply]
Oh, yes. I forgot to tell that you are not supposed to mess around with registers this any more since this is part of the regular API now.
i2s_config_t has fixed_mclk and i2s_pin_config_t has mck_io_num
